File: python/oop.py
# oop.py

# OOP in Python

# Class definition with Encapsulation
from abc import ABC, abstractmethod


class Animal:
    def __init__(self, name: str, species: str):
        self.name = name
        self._species = species  # Protected attribute

    def speak(self) -> str:
        raise NotImplementedError("Subclasses must implement this method")

    def get_species(self) -> str:
        """Public method to access the protected _species attribute."""
        return self._species

    def _private_method(self):
        """Private method (by convention, not enforced)."""
        return "This is a private method"


# Inheritance and Method Overriding
class Dog(Animal):
    def __init__(self, name: str, breed: str):
        super().__init__(name, "Dog")
        self.breed = breed

    def speak(self) -> str:
        return f"{self.name} says Woof!"

    def get_breed(self) -> str:
        """Public method to access the breed."""
        return self.breed


class Cat(Animal):
    def __init__(self, name: str, color: str):
        super().__init__(name, "Cat")
        self.color = color

    def speak(self) -> str:
        return f"{self.name} says Meow!"

    def get_color(self) -> str:
        """Public method to access the color."""
        return self.color


# Polymorphism
def make_animal_speak(animal: Animal):
    print(animal.speak())


dog = Dog("Buddy", "Golden Retriever")
cat = Cat("Whiskers", "Black")

make_animal_speak(dog)
make_animal_speak(cat)

# Abstract Base Classes


class Vehicle(ABC):
    def __init__(self, make: str, model: str):
        self.make = make
        self.model = model

    @abstractmethod
    def drive(self):
        """Abstract method that must be implemented by subclasses."""
        pass


class Car(Vehicle):
    def drive(self):
        return f"Driving a car: {self.make} {self.model}"


class Motorcycle(Vehicle):
    def drive(self):
        return f"Riding a motorcycle: {self.make} {self.model}"


car = Car("Toyota", "Corolla")
motorcycle = Motorcycle("Harley", "Davidson")

print(car.drive())
print(motorcycle.drive())


# Class Methods and Static Methods
class MathOperations:
    def __init__(self, value: int):
        self.value = value

    @classmethod
    def from_string(cls, value_str: str):
        """Alternative constructor using a class method."""
        return cls(int(value_str))

    @staticmethod
    def add(x: int, y: int) -> int:
        """Static method for adding two numbers."""
        return x + y


operation = MathOperations.from_string("10")
print(operation.value)  # 10
print(MathOperations.add(5, 3))  # 8


# Encapsulation and Property Decorators
class BankAccount:
    def __init__(self, balance: float):
        self._balance = balance

    @property
    def balance(self) -> float:
        """Property decorator for the balance."""
        return self._balance

    @balance.setter
    def balance(self, amount: float) -> None:
        if amount < 0:
            raise ValueError("Balance cannot be negative")
        self._balance = amount

    def deposit(self, amount: float) -> None:
        if amount > 0:
            self._balance += amount

    def withdraw(self, amount: float) -> None:
        if 0 < amount <= self._balance:
            self._balance -= amount
        else:
            raise ValueError("Invalid withdraw amount")


account = BankAccount(1000)
print(account.balance)  # 1000

account.deposit(500)
print(account.balance)  # 1500

account.withdraw(200)
print(account.balance)  # 1300

# Attempting to set a negative balance (will raise an exception)
try:
    account.balance = -100
except ValueError as e:
    print(e)

File: javascript/oop.js
// OOP in JavaScript

// 1. Class Definition and Constructor
class Animal {
  constructor(name) {
    this.name = name;
  }

  speak() {
    throw new Error("Subclasses must implement this method");
  }
}

// 2. Inheritance
class Dog extends Animal {
  speak() {
    return `${this.name} says Woof!`;
  }
}

class Cat extends Animal {
  speak() {
    return `${this.name} says Meow!`;
  }
}

// 3. Polymorphism
function makeAnimalSpeak(animal) {
  console.log(animal.speak());
}

const dog = new Dog("Buddy");
const cat = new Cat("Whiskers");

makeAnimalSpeak(dog); // Outputs: Buddy says Woof!
makeAnimalSpeak(cat); // Outputs: Whiskers says Meow!

// 4. Encapsulation
class User {
  #password; // Private property

  constructor(username, password) {
    this.username = username;
    this.#password = password;
  }

  checkPassword(password) {
    return this.#password === password;
  }

  setPassword(newPassword) {
    this.#password = newPassword;
  }
}

const user = new User("Alice", "secret");
console.log(user.checkPassword("secret")); // Outputs: true
console.log(user.checkPassword("wrong")); // Outputs: false

// 5. Static Methods and Properties
class MathUtils {
  static pi = 3.14159;

  static circumference(radius) {
    return 2 * MathUtils.pi * radius;
  }
}

console.log(
  "Circumference of circle with radius 5:",
  MathUtils.circumference(5),
); // Outputs: 31.4159

// 6. Getters and Setters
class Rectangle {
  constructor(width, height) {
    this._width = width;
    this._height = height;
  }

  get area() {
    return this._width * this._height;
  }

  set width(newWidth) {
    this._width = newWidth;
  }

  set height(newHeight) {
    this._height = newHeight;
  }
}

const rect = new Rectangle(10, 5);
console.log("Area of rectangle:", rect.area); // Outputs: 50
rect.width = 15;
rect.height = 10;
console.log("New area of rectangle:", rect.area); // Outputs: 150

// 7. Abstract Classes (ES6 does not support true abstract classes, but we can simulate them)
class Shape {
  constructor(name) {
    if (this.constructor === Shape) {
      throw new Error("Cannot instantiate abstract class Shape directly");
    }
    this.name = name;
  }

  area() {
    throw new Error("Method 'area()' must be implemented");
  }
}

class Circle extends Shape {
  constructor(radius) {
    super("Circle");
    this.radius = radius;
  }

  area() {
    return Math.PI * this.radius ** 2;
  }
}

const circle = new Circle(5);
console.log(`${circle.name} area:`, circle.area()); // Outputs: Circle area: 78.53981633974483

// 8. Inheritance with Method Overriding
class Bird extends Animal {
  constructor(name) {
    super(name);
  }

  speak() {
    return `${this.name} says Tweet!`;
  }
}

const bird = new Bird("Tweety");
makeAnimalSpeak(bird); // Outputs: Tweety says Tweet!

